spring-kafka/src/main/java/org/springframework/kafka/requestreply/AggregatingReplyingKafkaTemplate.java

```diff
@@ -0,0 +1,247 @@
+	 * @param producerFactory the producer factory.
+	 * @param replyContainer the reply container.
+	 * @param releaseStrategy the release strategy.
+	 */
+	public AggregatingReplyingKafkaTemplate(ProducerFactory<K, V> producerFactory,
+			GenericMessageListenerContainer<K, Collection<ConsumerRecord<K, R>>> replyContainer,
+			Predicate<Collection<ConsumerRecord<K, R>>> releaseStrategy) {
+
+		super(producerFactory, replyContainer);
+		Assert.notNull(releaseStrategy, "'releaseStrategy' cannot be null");
+		AckMode ackMode = replyContainer.getContainerProperties().getAckMode();
+		Assert.isTrue(ackMode.equals(AckMode.MANUAL) || ackMode.equals(AckMode.MANUAL_IMMEDIATE),
+				"The reply container must have a MANUAL or MANUAL_IMMEDIATE AckMode");
+		this.releaseStrategy = releaseStrategy;
+	}
+
+	/**
+	 * Set the timeout to use when committing offsets.
+	 * @param commitTimeout the timeout.
+	 */
+	public void setCommitTimeout(Duration commitTimeout) {
+		Assert.notNull(commitTimeout, "'commitTimeout' cannot be null");
+		this.commitTimeout = commitTimeout;
+	}
+
+	/**
+	 * Set to true to return a partial result when a request times out.
+	 * @param returnPartialOnTimeout true to return a partial result.
+	 */
+	public void setReturnPartialOnTimeout(boolean returnPartialOnTimeout) {
+		this.returnPartialOnTimeout = returnPartialOnTimeout;
+	}
+
+	@Override
+	public void onMessage(List<ConsumerRecord<K, Collection<ConsumerRecord<K, R>>>> data, Consumer<?, ?> consumer) {
+		List<ConsumerRecord<K, Collection<ConsumerRecord<K, R>>>> completed = new ArrayList<>();
+		data.forEach(record -> {
+			Header correlation = record.headers().lastHeader(KafkaHeaders.CORRELATION_ID);
+			if (correlation == null) {
+				this.logger.error("No correlationId found in reply: " + record
+						+ " - to use request/reply semantics, the responding server must return the correlation id "
+						+ " in the '" + KafkaHeaders.CORRELATION_ID + "' header");
+			}
+			else {
+				CorrelationKey correlationId = new CorrelationKey(correlation.value());
+				synchronized (this) {
+					if (isPending(correlationId)) {
+						List<ConsumerRecord<K, R>> list = addToCollection(record, correlationId).stream()
+								.map(entry -> entry.getRecord())
+								.collect(Collectors.toList());
+						if (this.releaseStrategy.test(list)) {
+							ConsumerRecord<K, Collection<ConsumerRecord<K, R>>> done =
+									new ConsumerRecord<>(AGGREGATED_RESULTS_TOPIC, 0, 0L, null, list);
+							done.headers()
+								.add(new RecordHeader(KafkaHeaders.CORRELATION_ID, correlationId.getCorrelationId()));
+							this.pending.remove(correlationId);
+							checkOffsetsAndCommitIfNecessary(list, consumer);
+							completed.add(done);
+						}
+					}
+					else {
+						logLateArrival(record, correlationId);
+					}
+				}
+			}
+		});
+		if (completed.size() > 0) {
+			super.onMessage(completed);
+		}
+	}
+
+	@Override
+	protected synchronized boolean handleTimeout(CorrelationKey correlationId,
```

Comment:
Why is this? 
Looks like such this method is called only from a single scheduled thread.

Reply:
`pending` needs to be protected.

Consider this race...

Thread in `onMessage()` calls `isPending(correlationId)` (true)
Timeout thread fires and removes correlationId from `pending`
`onMessage()` thread adds new delivery to `pending`.

Result: orphan